app/src/test/java/org/simple/clinic/summary/medicalhistory/MedicalHistorySummaryUiControllerTest.kt

```diff
@@ -0,0 +1,65 @@
+package org.simple.clinic.summary.medicalhistory
+
+import com.nhaarman.mockito_kotlin.doReturn
+import com.nhaarman.mockito_kotlin.mock
+import com.nhaarman.mockito_kotlin.verify
+import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
+import com.nhaarman.mockito_kotlin.whenever
+import io.reactivex.Observable
+import io.reactivex.disposables.Disposable
+import io.reactivex.subjects.PublishSubject
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.simple.clinic.medicalhistory.MedicalHistoryRepository
+import org.simple.clinic.patient.PatientMocker
+import org.simple.clinic.summary.PatientSummaryScreenUi
+import org.simple.clinic.widgets.ScreenCreated
+import org.simple.clinic.widgets.UiEvent
+import org.threeten.bp.Instant
+import java.util.UUID
+
+class MedicalHistorySummaryUiControllerTest {
+
+  private val patientUuid = UUID.fromString("31665de6-0265-4e33-888f-526bdb274699")
+
+  private val repository = mock<MedicalHistoryRepository>()
+  private val ui = mock<PatientSummaryScreenUi>()
+  private val medicalHistorySummaryUi = mock<MedicalHistorySummaryUi>()
+
+  private val events = PublishSubject.create<UiEvent>()
+
+  lateinit var controller: MedicalHistorySummaryUiController
+  lateinit var controllerSubscription: Disposable
+
+  @Before
+  fun setUp() {
+    whenever(ui.medicalHistorySummaryUi()) doReturn medicalHistorySummaryUi
+  }
+
+  @After
+  fun tearDown() {
+    controllerSubscription.dispose()
+  }
+
+  @Test
+  fun `patient's medical history should be populated`() {
+    // given
+    val medicalHistory = PatientMocker.medicalHistory(updatedAt = Instant.parse("2018-01-01T00:00:00Z"))
+    whenever(repository.historyForPatientOrDefault(patientUuid)) doReturn Observable.just(medicalHistory)
+
+    // when
+    setupController()
+
+    // then
+    verify(medicalHistorySummaryUi).populateMedicalHistory(medicalHistory)
+    verifyNoMoreInteractions(medicalHistorySummaryUi)
+  }
+
+  private fun setupController() {
+    controller = MedicalHistorySummaryUiController(patientUuid, repository)
+    controllerSubscription = events.compose(controller).subscribe { it.invoke(ui) }
+
+    events.onNext(ScreenCreated())
```

Comment:
I feel the events being triggered should be part of the tests themselves and not the setupController() method.
As the name of the method suggests, I would only expect some initialization to happen and not some events being sent.

Reply:
You are right about the `setupController` being method being expected to do only initialization, and no events being triggered here.

However, I feel that the `ScreenCreated` event is kind of a special case. The only reason it exists is because our v1 architecture does not have any way to be notified when the view is "ready" since it is modelled on a stream. The only reason that `ScreenCreated` exists is so that we can send this to the stream as a "trigger".

If we consider the Mobius architecture, we do not need to use `ScreenCreated` here because that job is being done by the `Init` function. Regardless of whether screen is written in Mobius or the architecture v1, the (controller) tests for them will end up having a general structure like so:


```kotlin
fun test() {
  // given
  // set-up mocks here

  // when
  // simulate screen set-up here
  // trigger user events

  // then
  // verify
}
```

If we move the emission of the `ScreenCreated` event outside the controller setup and into the `trigger user events` section, then we will have to continuously remove this line for every test that is migrated to Mobius. This is an indication that it is not truly a "user" event but something that is specific to the v1 controller.

Does this make sense?

Reply:
Given that we are not re-writing tests for now. We are still going to continue reading the existing `LogicTests`. I agree that `ScreenCreated` is not relevant for Mobius architecture. But given that these tests are not written in the Mobius format either, I think it is ok to be explicit that this event is being triggered after the controller is setup. That is how we are creating this event in the older architecture. 
My point is that this is sort of a hidden and confusing behavior for someone reading the tests. Also why do we need to setup the controller inside every test? Is it because of the event?
We can extract this to the `setup` method of of the test suite since we are testing the controller, it is safe to assume we would need it to be setup for every test. Any specific reason for doing so?

Reply:
> I agree that ScreenCreated is not relevant for Mobius architecture. But given that these tests are not written in the Mobius format either, I think it is ok to be explicit that this event is being triggered after the controller is setup. That is how we are creating this event in the older architecture.

I forgot to mention that `ScreenCreated` will end up getting removed from the older style controller tests when we migrate the controller to Mobius. If you look at `BloodPressureEntrySheetLogicTest`, the controller was migrated to Mobius. We no longer have any `ScreenCreated` events here.

> Also why do we need to setup the controller inside every test?
We can extract this to the setup method of of the test suite since we are testing the controller, it is safe to assume we would need it to be setup for every test. Any specific reason for doing so?

Yes. Our current style of setting up controllers in the `@Before` fixture does not allow us to setup stubs before the controller is initialized. Since the controller subscribes to the streams before the mocks are setup in many cases, this makes testing in a lot of cases hard and we need to jump through hoops (like using a `PublishSubject`, for instance) to setup the test correctly. This is why as a matter of policy, we will no longer setup the controller in the `@Before` fixture and will manually do this in every test before refactoring to Mobius.

If you look at all the older controller tests where they've been moved to Mobius, we've done this all the time to make the refactoring and testing easier.

These are some of the older style tests that we have migrated:
- `BloodPressureEntrySheetLogicTest`
- `PatientEntryScreenLogicTest`
- `AllPatientsInFaciltyLogicTest`

Reply:
So the reason for using this method and calling it in each test is that in the future once we move this controller to Mobius, we would need to do instantiate the controller in each test?

Reply:
Yup, we'll do this in future PRs.